actors/migration/nv15/miner.go

```diff
@@ -19,63 +25,184 @@ type minerMigrator struct{}
 func (m minerMigrator) migrateState(ctx context.Context, store cbor.IpldStore, in actorMigrationInput) (*actorMigrationResult, error) {
 	var inState miner6.State
 	if err := store.Get(ctx, in.head, &inState); err != nil {
-		return nil, err
+		return nil, xerrors.Errorf("getting inState: %w", err)
 	}
 
-	sectorsOut, err := migrateSectors(ctx, store, inState.Sectors)
+	outState := fromv6State(inState)
+	ctxStore := adt.WrapStore(ctx, store)
+
+	sectorsOut, err := migrateSectors(ctx, ctxStore, in.cache, in.address, inState.Sectors)
 	if err != nil {
-		return nil, err
+		return nil, xerrors.Errorf("failed to migrate sectors: %w", err)
 	}
 
-	outState := miner7.State{
-		Info:                       inState.Info,
-		PreCommitDeposits:          inState.PreCommitDeposits,
-		LockedFunds:                inState.LockedFunds,
-		VestingFunds:               inState.VestingFunds,
-		FeeDebt:                    inState.FeeDebt,
-		InitialPledge:              inState.InitialPledge,
-		PreCommittedSectors:        inState.PreCommittedSectors,
-		PreCommittedSectorsCleanUp: inState.PreCommittedSectorsCleanUp,
-		AllocatedSectors:           inState.AllocatedSectors,
-		Sectors:                    sectorsOut,
-		ProvingPeriodStart:         inState.ProvingPeriodStart,
-		CurrentDeadline:            inState.CurrentDeadline,
-		Deadlines:                  inState.Deadlines,
-		EarlyTerminations:          inState.EarlyTerminations,
-		DeadlineCronActive:         inState.DeadlineCronActive,
+	outState.Sectors = sectorsOut
+
+	deadlinesOut, err := m.migrateDeadlines(ctx, ctxStore, in.cache, inState.Deadlines, sectorsOut)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to migrate deadlines: %w", err)
 	}
 
+	outState.Deadlines = deadlinesOut
+
 	newHead, err := store.Put(ctx, &outState)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to flush outState: %w", err)
+	}
+
 	return &actorMigrationResult{
 		newCodeCID: m.migratedCodeCID(),
 		newHead:    newHead,
-	}, err
+	}, nil
 }
 
 func (m minerMigrator) migratedCodeCID() cid.Cid {
 	return builtin7.StorageMinerActorCodeID
 }
 
-func migrateSectors(ctx context.Context, store cbor.IpldStore, inRoot cid.Cid) (cid.Cid, error) {
-	ctxStore := adt.WrapStore(ctx, store)
-	inArray, err := adt.AsArray(ctxStore, inRoot, miner6.SectorsAmtBitwidth)
+func migrateSectors(ctx context.Context, store adt.Store, cache MigrationCache, minerAddr address.Address, inRoot cid.Cid) (cid.Cid, error) {
+	return cache.Load(CidKey(inRoot), func() (cid.Cid, error) {
```

Comment:
The overwhelming majority of actors should be covered in the cache here:

- Even if this is the first premigration, the majority of miners are "empty" and so have identical sectors AMT roots
- If a premigration HAS already run, the non-empty miners will very likely be covered here because the sectors AMT doesn't change that much -- new sectors and extensions are the only things that can change it, i think (._..and CompactPartitions_)

Reply:
For peace of mind use the pattern we established last year to type cache entries: 
1. Types: https://github.com/filecoin-project/specs-actors/blob/a2aec9c6af0e55ff9f7da257b9b5ca4c3e88b8e9/actors/migration/nv10/top.go#L47
2. Usage: https://github.com/filecoin-project/specs-actors/blob/release/v3/actors/migration/nv10/miner.go#L40

Data collisions (the thing being hashed as an AMT somehow has the same bytes as the thing being hashed as miner state) really should be impossible but it would be nice not to think about this at all.

Reply:
Yessir!